File: 1114 - Print in Order.cpp
#include <vector>
#include <mutex>
#include <semaphore.h>

using namespace std;

class Foo {
private:
    sem_t a, b;

public:
    Foo(){
        sem_init(&a, 0, 0);
        sem_init(&b, 0, 0);
    }
    

    void first(function<void()> printFirst) {

        // printFirst() outputs "first". Do not change or remove this line.

        printFirst();
        sem_post(&a);
    }

    void second(function<void()> printSecond) {

        // printSecond() outputs "second". Do not change or remove this line.
        sem_wait(&a);
        printSecond();
        sem_post(&b);
    }

    void third(function<void()> printThird) {

        // printThird() outputs "third". Do not change or remove this line.
        sem_wait(&b);
        printThird();

    }
};

File: 1114 - Print in Order.java
class Foo {
    private boolean a = false, b = false;
    public Foo() {
        a = false;
        b = false;
    }

    public void first(Runnable printFirst) throws InterruptedException {

        // printFirst.run() outputs "first". Do not change or remove this line.
        printFirst.run();
        a = true;
    }

    public void second(Runnable printSecond) throws InterruptedException {

        // printSecond.run() outputs "second". Do not change or remove this line.
        while (!a) continue;
        printSecond.run();
        b = true;
    }

    public void third(Runnable printThird) throws InterruptedException {

        // printThird.run() outputs "third". Do not change or remove this line.
        while (!a || !b) continue;
        printThird.run();

    }
}